### ISSUE TITLE
Close upstream connections on downstream disconnect

### ISSUE THREAD

**Issue by howardjohn** [istio/ztunnel#347]
I think right now we still hold upstream connections when downstream disconnects. At the very least, we do when we are trying to connect to the upstream; at some point we must disconnect (probably when calling `read`). We may want to proactively stop upstream connections. Without this, I think there is some DOS ability where a downstream opens+closes a bunch of connections, causing the ztunnel to accumulate many connections until the connection timeout

**Comment by ilrudie** [comment]
Is this still needed after #293? The pool has idle timeout and being able to reuse the connection should prevent many runaway connection accumulation issues.

**Comment by bleggett** [comment]
> The pool is the outer connections, the inner connections are still accumulated though. This may actually impact applications as they will not be aware the peer closed the connection?

It's ~probably~ fine for most things in terms of disruption, but it's not desirable due to the addl latency it might create. 

In general we should probably signal conndrop to downstream as soon as we get it for upstream - delay there is us adding latency, effectively.

**Comment by hzxuzhonghu** [comment]
Not sure if this is right, for inbound processing why donot we close both stream when proxy::relay returns

**Comment by jaellio** [comment]
@howardjohn has this issue been resolved?

**Comment by howardjohn** [comment]
No, I don't think it has been. We cannot curerntly detect when downstream disconnects

**Comment by ilrudie** [comment]
Is this an enhancement we'll need from hyper then? Maybe a way to register a function to be called on disconnect or similar?

**Comment by howardjohn** [comment]
Hyper doesn't manage the connection, this would be tokio level I think

**Comment by ilrudie** [comment]
I was thinking something like hyper::server::conn::http2::Connection could potentially have this functionality. Maybe it would need to be tokio TcpStream though 🤔

**Comment by ilrudie** [comment]
Just brainstorming. Could we be using poll_peak looking for connection rst/abort IO errors or similar? I think it's a little kludgy and being able to register something like an on_disconnect function would be nicer but maybe it's something we could pursue more independently.
